### ISSUE TITLE
Scan iterator fix for compilation under Visual Studio 2010

### ISSUE THREAD

**Issue by nnposter** [nmap/nmap#560]
According to published [Windows source code compilation instructions](https://nmap.org/book/inst-windows.html#inst-win-source) users should be able to compile Nmap with Visual Studio 2010. Unfortunately this is not possible as of r36088 (7c0280382e8682c151e73cd2afa560d304603462).

Namely the compilation is failing with errors like:
```
..\scan_engine.cc(765): error C2679: binary '=' : no operator found which takes a right-hand operand of type 'std::_Tree_const_iterator<_Mytree>' (or there is no acceptable conversion)
          with
          [
              _Mytree=std::_Tree_val<std::_Tset_traits<HostScanStats *,HssPredicate,std::allocator<HostScanStats *>,false>>
          ]
          c:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\include\xtree(337): could be 'std::_Tree_const_iterator<_Mytree> &std::_Tree_const_iterator<_Mytree>::operator =(const std::_Tree_const_iterator<_Mytree> &)'
          with
          [
              _Mytree=std::_Tree_val<std::_Tset_traits<HostScanStats *,std::less<HostScanStats *>,std::allocator<HostScanStats *>,false>>
          ]
          while trying to match the argument list '(std::_Tree_const_iterator<_Mytree>, std::_Tree_const_iterator<_Mytree>)'
          with
          [
              _Mytree=std::_Tree_val<std::_Tset_traits<HostScanStats *,std::less<HostScanStats *>,std::allocator<HostScanStats *>,false>>
          ]
          and
          [
              _Mytree=std::_Tree_val<std::_Tset_traits<HostScanStats *,HssPredicate,std::allocator<HostScanStats *>,false>>
          ]
```
The following patch has resolved the issue for me but it warrants scrutiny from others before committing.
```
* Fixes scan iterator definitions to make the code compilable with VS2010
--- a/scan_engine.cc
+++ b/scan_engine.cc
@@ -760,7 +760,7 @@
 }
 
 UltraScanInfo::~UltraScanInfo() {
-  std::set<HostScanStats *>::iterator hostI;
+  std::set<HostScanStats *, HssPredicate>::iterator hostI;
 
   for (hostI = incompleteHosts.begin(); hostI != incompleteHosts.end(); hostI++) {
     delete *hostI;
@@ -819,7 +819,7 @@
 /* Return a number between 0.0 and 1.0 inclusive indicating how much of the scan
    is done. */
 double UltraScanInfo::getCompletionFraction() {
-  std::set<HostScanStats *>::iterator hostI;
+  std::set<HostScanStats *, HssPredicate>::iterator hostI;
   double total;
 
   /* Add 1 for each completed host. */
@@ -1076,7 +1076,7 @@
 bool UltraScanInfo::sendOK(struct timeval *when) {
   struct timeval lowhtime = {0};
   struct timeval tmptv;
-  std::set<HostScanStats *>::iterator host;
+  std::set<HostScanStats *, HssPredicate>::iterator host;
   bool ggood = false;
   bool thisHostGood = false;
   bool foundgood = false;
@@ -1134,7 +1134,7 @@
 /* Find a HostScanStats by its IP address in the incomplete and completed lists.
    Returns NULL if none are found. */
 HostScanStats *UltraScanInfo::findHost(struct sockaddr_storage *ss) {
-  std::set<HostScanStats *>::iterator hss;
+  std::set<HostScanStats *, HssPredicate>::iterator hss;
 
   HssPredicate::ss = ss;
   HostScanStats *fakeHss = NULL;
@@ -1160,7 +1160,7 @@
    is here to replace numIncompleteHosts() < n, which would have to walk
    through the entire list. */
 bool UltraScanInfo::numIncompleteHostsLessThan(unsigned int n) {
-  std::set<HostScanStats *>::iterator hostI;
+  std::set<HostScanStats *, HssPredicate>::iterator hostI;
   unsigned int count;
 
   count = 0;
@@ -1180,7 +1180,7 @@
    list, and remove any hosts from completedHosts which have exceeded their
    lifetime.  Returns the number of hosts removed. */
 int UltraScanInfo::removeCompletedHosts() {
-  std::set<HostScanStats *>::iterator hostI, nxt;
+  std::set<HostScanStats *, HssPredicate>::iterator hostI, nxt;
   HostScanStats *hss = NULL;
   int hostsRemoved = 0;
   bool timedout = false;
@@ -2290,7 +2290,7 @@
 }
 
 static void doAnyPings(UltraScanInfo *USI) {
-  std::set<HostScanStats *>::iterator hostI;
+  std::set<HostScanStats *, HssPredicate>::iterator hostI;
   HostScanStats *hss = NULL;
 
   gettimeofday(&USI->now, NULL);
@@ -2365,7 +2365,7 @@
 /* Go through the ProbeQueue of each host, identify any
    timed out probes, then try to retransmit them as appropriate */
 static void doAnyOutstandingRetransmits(UltraScanInfo *USI) {
-  std::set<HostScanStats *>::iterator hostI;
+  std::set<HostScanStats *, HssPredicate>::iterator hostI;
   std::list<UltraProbe *>::iterator probeI;
   /* A cache of the last processed probe from each host, to avoid re-examining a
      bunch of probes to find the next one that needs to be retransmitted. */
@@ -2445,7 +2445,7 @@
 /* Print occasional remaining time estimates, as well as
    debugging information */
 static void printAnyStats(UltraScanInfo *USI) {
-  std::set<HostScanStats *>::iterator hostI;
+  std::set<HostScanStats *, HssPredicate>::iterator hostI;
   HostScanStats *hss;
   struct ultra_timing_vals hosttm;
 
@@ -2515,7 +2515,7 @@
 /* Go through the data structures, making appropriate changes (such as expiring
    probes, noting when hosts are complete, etc. */
 static void processData(UltraScanInfo *USI) {
-  std::set<HostScanStats *>::iterator hostI;
+  std::set<HostScanStats *, HssPredicate>::iterator hostI;
   std::list<UltraProbe *>::iterator probeI, nextProbeI;
   HostScanStats *host = NULL;
   UltraProbe *probe = NULL;
--- a/scan_engine.h
+++ b/scan_engine.h
@@ -759,7 +759,7 @@
 private:
 
   unsigned int numInitialTargets;
-  std::set<HostScanStats *>::iterator nextI;
+  std::set<HostScanStats *, HssPredicate>::iterator nextI;
 
 };
 
--- a/scan_engine_connect.cc
+++ b/scan_engine_connect.cc
@@ -479,7 +479,7 @@
   int timeleft;
   ConnectScanInfo *CSI = USI->gstats->CSI;
   int sd;
-  std::set<HostScanStats *>::iterator hostI;
+  std::set<HostScanStats *, HssPredicate>::iterator hostI;
   HostScanStats *host;
   UltraProbe *probe = NULL;
   int optval;
@@ -520,7 +520,7 @@
      and find the relevant ones. Note the peculiar structure of the loop--we
      iterate through both incompleteHosts and completedHosts, because global
      timing pings are sent to hosts in completedHosts. */
-  std::set<HostScanStats *>::iterator incompleteHostI, completedHostI;
+  std::set<HostScanStats *, HssPredicate>::iterator incompleteHostI, completedHostI;
   incompleteHostI = USI->incompleteHosts.begin();
   completedHostI = USI->completedHosts.begin();
   while ((incompleteHostI != USI->incompleteHosts.end()
```

**Comment by hsluoyz** [comment]
Hi @nnposter ,

I think that statement is already out-of-date? In the same page it also said:

> Due to the way Nmap is compiled, it requires the Microsoft Visual C++ 2013 Redistributable Package of runtime components.

And the ``nmap.sln`` clearly states:
```
Microsoft Visual Studio Solution File, Format Version 12.00
# Visual Studio 2013
VisualStudioVersion = 12.0.21005.1
```

So Nmap is currently developed by Visual Studio 2013.

And according to Microsoft [here](https://support.microsoft.com/en-us/lifecycle/search?alpha=visual%20studio%202010), Visual Studio 2010 should have ended its support since ``6/29/2010``. Even Visual Studio 2013 will end its support in only a few days (``10/11/2016``). So I believe we can just deprecate the VS 2010 compiling way on Nmap and move on to VS 2013 and VS 2015 completely.

**Comment by nnposter** [comment]
IMHO the first two arguments do not count while the third one is incorrect. 

The first one merely states that because the official binaries were produced with VS2013 then the users need to have the corresponding redistributable libraries. The statement is irrelevant to those users that compile Nmap by themselves.

The second one is similarly irrelevant. Just because the included convenience VS files were generated with VS2013 does not mean that VS2013 is needed for the compilation.

In your third argument the stated VS2010 date, 6/29/2010, is the *start* date, not end date. The extended support will continue for almost four more years, till 2020. Similarly, the end of support for VS2013 is not 2016. Mainstream support ends in 2019 and extended in 2024.

In summary, there might exist reasons for dropping VS2010 compatibility but I do not believe that you have presented any tangible argument why the issue should not be fixed.

**Comment by hsluoyz** [comment]
> The first one merely states that because the official binaries were produced with VS2013 then the users need to have the corresponding redistributable libraries. The statement is irrelevant to those users that compile Nmap by themselves.

> The second one is similarly irrelevant. Just because the included convenience VS files were generated with VS2013 does not mean that VS2013 is needed for the compilation.

I'm just saying that the "recommended" IDE is VS2013. It's certainly good if you can make VS2010 compile too. This fix seems to be feasible for me.

> In your third argument the stated VS2010 date, 6/29/2010, is the start date, not end date. The extended support will continue for almost four more years, till 2020. Similarly, the end of support for VS2013 is not 2016. Mainstream support ends in 2019 and extended in 2024.

I misunderstood it.

**Comment by nnposter** [comment]
@dmiller-nmap  I saw the changes in r36365 earlier today and noted the implied termination of support for VS2010. Is there a reason to do so? The only compilation issue I am aware of is this one.

**Comment by Tudor-Coman** [comment]
Hi,

I'm responsible for adding that code, I wasn't aware it caused problems on VS2010.
I just recently modified that std::set to a std::multiset to solve a separate issue so the fix should look like this:

```
Index: scan_engine.cc
===================================================================
--- scan_engine.cc      (revision 36365)
+++ scan_engine.cc      (working copy)
@@ -760,7 +760,7 @@
 }

 UltraScanInfo::~UltraScanInfo() {
-  std::multiset<HostScanStats *>::iterator hostI;
+  std::multiset<HostScanStats *, HssPredicate>::iterator hostI;

   for (hostI = incompleteHosts.begin(); hostI != incompleteHosts.end(); hostI++) {
     delete *hostI;
@@ -819,7 +819,7 @@
 /* Return a number between 0.0 and 1.0 inclusive indicating how much of the scan
    is done. */
 double UltraScanInfo::getCompletionFraction() {
-  std::multiset<HostScanStats *>::iterator hostI;
+  std::multiset<HostScanStats *, HssPredicate>::iterator hostI;
   double total;

   /* Add 1 for each completed host. */
@@ -1076,7 +1076,7 @@
 bool UltraScanInfo::sendOK(struct timeval *when) {
   struct timeval lowhtime = {0};
   struct timeval tmptv;
-  std::multiset<HostScanStats *>::iterator host;
+  std::multiset<HostScanStats *, HssPredicate>::iterator host;
   bool ggood = false;
   bool thisHostGood = false;
   bool foundgood = false;
@@ -1134,7 +1134,7 @@
 /* Find a HostScanStats by its IP address in the incomplete and completed lists.
    Returns NULL if none are found. */
 HostScanStats *UltraScanInfo::findHost(struct sockaddr_storage *ss) {
-  std::multiset<HostScanStats *>::iterator hss;
+  std::multiset<HostScanStats *, HssPredicate>::iterator hss;

   HssPredicate::ss = ss;
   HostScanStats *fakeHss = NULL;
@@ -1160,7 +1160,7 @@
    is here to replace numIncompleteHosts() < n, which would have to walk
    through the entire list. */
 bool UltraScanInfo::numIncompleteHostsLessThan(unsigned int n) {
-  std::multiset<HostScanStats *>::iterator hostI;
+  std::multiset<HostScanStats *, HssPredicate>::iterator hostI;
   unsigned int count;

   count = 0;
@@ -1180,7 +1180,7 @@
    list, and remove any hosts from completedHosts which have exceeded their
    lifetime.  Returns the number of hosts removed. */
 int UltraScanInfo::removeCompletedHosts() {
-  std::multiset<HostScanStats *>::iterator hostI, nxt;
+  std::multiset<HostScanStats *, HssPredicate>::iterator hostI, nxt;
   HostScanStats *hss = NULL;
   int hostsRemoved = 0;
   bool timedout = false;
@@ -2290,7 +2290,7 @@
 }

 static void doAnyPings(UltraScanInfo *USI) {
-  std::multiset<HostScanStats *>::iterator hostI;
+  std::multiset<HostScanStats *, HssPredicate>::iterator hostI;
   HostScanStats *hss = NULL;

   gettimeofday(&USI->now, NULL);
@@ -2365,7 +2365,7 @@
 /* Go through the ProbeQueue of each host, identify any
    timed out probes, then try to retransmit them as appropriate */
 static void doAnyOutstandingRetransmits(UltraScanInfo *USI) {
-  std::multiset<HostScanStats *>::iterator hostI;
+  std::multiset<HostScanStats *, HssPredicate>::iterator hostI;
   std::list<UltraProbe *>::iterator probeI;
   /* A cache of the last processed probe from each host, to avoid re-examining a
      bunch of probes to find the next one that needs to be retransmitted. */
@@ -2445,7 +2445,7 @@
 /* Print occasional remaining time estimates, as well as
    debugging information */
 static void printAnyStats(UltraScanInfo *USI) {
-  std::multiset<HostScanStats *>::iterator hostI;
+  std::multiset<HostScanStats *, HssPredicate>::iterator hostI;
   HostScanStats *hss;
   struct ultra_timing_vals hosttm;

@@ -2515,7 +2515,7 @@
 /* Go through the data structures, making appropriate changes (such as expiring
    probes, noting when hosts are complete, etc. */
 static void processData(UltraScanInfo *USI) {
-  std::multiset<HostScanStats *>::iterator hostI;
+  std::multiset<HostScanStats *, HssPredicate>::iterator hostI;
   std::list<UltraProbe *>::iterator probeI, nextProbeI;
   HostScanStats *host = NULL;
   UltraProbe *probe = NULL;
Index: scan_engine.h
===================================================================
--- scan_engine.h       (revision 36365)
+++ scan_engine.h       (working copy)
@@ -759,7 +759,7 @@
 private:

   unsigned int numInitialTargets;
-  std::multiset<HostScanStats *>::iterator nextI;
+  std::multiset<HostScanStats *, HssPredicate>::iterator nextI;

 };

Index: scan_engine_connect.cc
===================================================================
--- scan_engine_connect.cc      (revision 36365)
+++ scan_engine_connect.cc      (working copy)
@@ -489,7 +489,7 @@
   int timeleft;
   ConnectScanInfo *CSI = USI->gstats->CSI;
   int sd;
-  std::multiset<HostScanStats *>::iterator hostI;
+  std::multiset<HostScanStats *, HssPredicate>::iterator hostI;
   HostScanStats *host;
   UltraProbe *probe = NULL;
   int optval;
@@ -530,7 +530,7 @@
      and find the relevant ones. Note the peculiar structure of the loop--we
      iterate through both incompleteHosts and completedHosts, because global
      timing pings are sent to hosts in completedHosts. */
-  std::multiset<HostScanStats *>::iterator incompleteHostI, completedHostI;
+  std::multiset<HostScanStats *, HssPredicate>::iterator incompleteHostI, completedHostI;
   incompleteHostI = USI->incompleteHosts.begin();
   completedHostI = USI->completedHosts.begin();
   while ((incompleteHostI != USI->incompleteHosts.end()

```

Currently I still have problems compiling because of the undeclared identifier ENONET in scan_engine_connect.cc but if I comment that, it compiles successfully.


Should I commit this?

**Comment by nnposter** [comment]
My vote would be "Yes".

On the broader issue of Nmap official compatibility with VS2010 my suggestion is to stop supporting it once there is a substantial benefit in doing so, not just because the official Windows build is done using VS2013.

**Comment by dmiller-nmap** [comment]
@Tudor-Coman Please do commit that. I'm hoping @W0naN0w will be committing the fix to the ENOENT issue soon via #568 so no worries.

@nnposter I don't intend to remove VS2010 support, so please continue to let us know when something breaks. I updated the documentation to indicate the version we are using to build, because that's what we can be sure works. The wording is "Compilation requires Microsoft Visual C++ 2013," but we can support older or newer versions as well. I'll look over the lifecycle docs and have a chat with Fyodor about a support plan, since it seems it would be best to document that somewhere.

**Comment by nnposter** [comment]
I can confirm that as of r36376 (0a7a7e395b2fd1e8ccfc8e9f70e6870feae69609) Nmap compiles without errors on VS2010.

As a side note, there are two unrelated areas causing compiler warnings but in both cases the issue is not serious and the fix would not be trivial so I am going to leave them as-is.
